### PR TITLE
torch.dropout: avoid unnecessary mul for p=1

### PR DESCRIPTION
**Pull request by kshitij12345**
Not sure how often anyone uses p=1 if at all but no mul with 0 is needed in that case.

Benchmark Script

```python
import torch
import itertools
import time
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import sys

print('Using pytorch %s' % (torch.__version__))

shapes = [(32,), (32, 32), (2, 16, 32), (2, 16, 32, 32), (8, 16, 64, 64)]
num_threads = 8
dtypes = [torch.float, torch.double]
repeats = 10

for dtype in dtypes:
    results = []
    for mat1_shape in shapes:
        mat1 = torch.rand(*mat1_shape, dtype=dtype, device='cpu')
        mat1_cuda = mat1.to('cuda')

        tasks = [("torch.nn.functional.dropout(mat1, p=1.)", "torch.dropout CPU"),
                 ("torch.nn.functional.dropout(mat1, p=1., inplace=True)", "torch.dropout inplace CPU"),
                 # In CUDA, this path is taken only when training is False.
                 ("torch.nn.functional.dropout(mat1_cuda, p=1., training=False, inplace=True)", "torch.dropout CUDA")]

        timers = [Timer(stmt=stmt, num_threads=num_threads, label=f"dropout {dtype}", sub_label=f"{(mat1_shape)}", description=label, globals=globals()) for stmt, label in tasks]

        for i, timer in enumerate(timers * repeats):
            results.append(
                timer.blocked_autorange()
            )
            print(f"\r{i + 1} / {len(timers) * repeats}", end="")
            sys.stdout.flush()
    comparison = Compare(results)
    comparison.print()
```

Note that this also affect, `feature_dropout`, `alpha_dropout`, `feature_alpha_dropout` their benchmark would follow the results below (as they use the same path for the given case).

Benchmark Before Change
```
Using pytorch 1.8.0a0+3f9697b
30 / 30[----------------------------------- dropout torch.float32 ------------------------------------]
                       |  torch.dropout CPU  |  torch.dropout inplace CPU  |  torch.dropout CUDA
8 threads: -------------------------------------------------------------------------------------
      (32,)            |          5.9        |              4.8            |         2.0        
      (32, 32)         |          6.2        |              4.9            |         2.0        
      (2, 16, 32)      |          6.3        |              5.0            |         2.0        
      (2, 16, 32, 32)  |         10.9        |              8.1            |         2.1        
      (8, 16, 64, 64)  |         27.3        |             20.1            |         2.1        

Times are in microseconds (us).

30 / 30[----------------------------------- dropout torch.float64 ------------------------------------]
                       |  torch.dropout CPU  |  torch.dropout inplace CPU  |  torch.dropout CUDA
8 threads: -------------------------------------------------------------------------------------
      (32,)            |          6.0        |              4.9            |         2.0        
      (32, 32)         |          6.5        |              5.2            |         2.0        
      (2, 16, 32)      |          6.5        |              5.3            |         2.1        
      (2, 16, 32, 32)  |         13.9        |             11.2            |         2.0        
      (8, 16, 64, 64)  |         39.2        |             33.2            |         2.1        

Times are in microseconds (us).
```

Benchmark After Change
```
Using pytorch 1.8.0a0+3f9697b
30 / 30[----------------------------------- dropout torch.float32 ------------------------------------]
                       |  torch.dropout CPU  |  torch.dropout inplace CPU  |  torch.dropout CUDA
8 threads: -------------------------------------------------------------------------------------
      (32,)            |          4.7        |              2.7            |         2.1        
      (32, 32)         |          4.8        |              2.8            |         2.1        
      (2, 16, 32)      |          4.9        |              2.8            |         2.1        
      (2, 16, 32, 32)  |          7.7        |              5.3            |         2.1        
      (8, 16, 64, 64)  |         17.3        |             13.8            |         2.1        

Times are in microseconds (us).

30 / 30[----------------------------------- dropout torch.float64 ------------------------------------]
                       |  torch.dropout CPU  |  torch.dropout inplace CPU  |  torch.dropout CUDA
8 threads: -------------------------------------------------------------------------------------
      (32,)            |          4.7        |              2.6            |         2.1        
      (32, 32)         |          5.0        |              2.9            |         2.1        
      (2, 16, 32)      |          4.9        |              2.9            |         2.1        
      (2, 16, 32, 32)  |         10.7        |              8.0            |         2.1        
      (8, 16, 64, 64)  |         25.6        |             22.5            |         2.1        

Times are in microseconds (us).
```

Reference `cuda` skips this path when training is True

https://github.com/pytorch/pytorch/blob/f90da88d8f53a318c2d4d6e60ad949df704c5273/aten/src/ATen/native/Dropout.cpp#L85-L95

https://github.com/pytorch/pytorch/blob/f90da88d8f53a318c2d4d6e60ad949df704c5273/aten/src/ATen/native/Dropout.cpp#L24-L26

### CODE REVIEW

**Comment by vadimkantorov** [comment]
One problem is that it changes the behavior from copying to no-copying. This can make some problems if there's an inplace op just after.

**Comment by kshitij12345** [comment]
@vadimkantorov 
> One problem is that it changes the behavior from copying to no-copying. This can make some problems if there's an inplace op just after only in train time, but maybe it's a minor consideration

Could you elaborate, cause I think the behaviour of inplace and copy would be preserved.

**Comment by heitorschueroff** [comment]
The only thing I'm concerned is that replacing `multiply` with `zero_tensor` will lose the sign on the zeros. The sign of zeros could change the result of atan2 for instance. @mruberry Can you advise if we want to preserve the sign of zeros here?

**Comment by heitorschueroff** [comment]
@kshitij12345 You are not synchronizing CUDA in your benchmark which is why you see the same results. You can refer to this [WIP benchmark recipe](https://deploy-preview-1176--pytorch-tutorials-preview.netlify.app/recipes/recipes/benchmark.html?highlight=benchmark) for a guide on benchmarking PyTorch code.

**Comment by ngimel** [comment]
`Timer` synchronizes cuda automatically, that said, 2us is extremely suspect.

**Comment by kshitij12345** [comment]
> The only thing I'm concerned is that replacing multiply with zero_tensor will lose the sign on the zeros. The sign of zeros could change the result of atan2 for instance. @mruberry Can you advise if we want to preserve the sign of zeros here?

Sample Snippet:
```python
>>> x = torch.randn(3,3)
>>> x
tensor([[ 2.2820, -0.9474,  0.1627],
        [ 1.0592,  0.4148, -0.5017],
        [-1.0321,  1.0442, -0.3100]])
>>> x * torch.zeros_like(x)
tensor([[0., -0., 0.],
        [0., 0., -0.],
        [-0., 0., -0.]])
>>> torch.zeros_like(x)
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])

```

**Comment by ngimel** [comment]
tbh, I don't think this corner case deserves special treatment, so I'd be against merging this PR. However, there are a couple of real dropout issues that need fixing
1) with p=0 we still go to special cuda kernel instead of taking a shortcut
2) in the cuda kernel, both vectorized and regular, the comparisons are `<`, whereas given that curand produces 0-exclusive, 1-inclusive range they should be `<=`. 
Can you refocus this PR to fix these issues please?

**Comment by kshitij12345** [comment]
> with p=0 we still go to special cuda kernel instead of taking a shortcut

This I believe is handled properly, fused kernel is taken only when `p > 0 and p < 1`,
https://github.com/pytorch/pytorch/blob/fe7d1d7d0e0762cca6b323f6e2ad7cc252f8ccf3/aten/src/ATen/native/Dropout.cpp#L24-L26

Otherwise input is returned directly if `p==0`,
https://github.com/pytorch/pytorch/blob/fe7d1d7d0e0762cca6b323f6e2ad7cc252f8ccf3/aten/src/ATen/native/Dropout.cpp#L44-L47

For point 2,
> in the cuda kernel, both vectorized and regular, the comparisons are <, whereas given that curand produces 0-exclusive, 1-inclusive range they should be <=

Logically it makes sense, but with above condition `p > 0 and p < 1`, `p==1` never reaches those kernel.

**Comment by ngimel** [comment]
Yeah, my mistake, turns out that for p=0 we do take the shortcut and thus no values should be dropped. Changing < to <= still makes sense even for probabilities in (0,1), but that's fixing a very small skew, so maybe not worth the effort.
